### PR TITLE
Fix headless service access in strict mode when autoMTLS enabled

### LINKED ISSUE #20200: Pods in same controllers should speak to each others in STRICT mode
Hello,

I just figured that pods in same controllers (i.e. deployment, sts, daemonset, etc...) can't speak to each other in mtls STRICT Mode.
Why is that ? Databases or kv engines need to speak to each others on their pod address for quorum or replication and that makes them unable to.

Could you clarify ?

Thanks,

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Hopefully to fix #20200

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@howardjohn @incfly

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -1194,21 +1194,26 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.TLSSetting
...
-		cluster.TransportSocketMatches = []*apiv2.Cluster_TransportSocketMatch{
-			{
-				Name: "tlsMode-" + model.IstioMutualTLSModeLabel,
-				Match: &structpb.Struct{
-					Fields: map[string]*structpb.Value{
-						model.TLSModeLabelShortname: {Kind: &structpb.Value_StringValue{StringValue: model.IstioMutualTLSModeLabel}},
+
+	// For `Cluster_ORIGINAL_DST` cluster discovery type, the LbPolicy is `Cluster_CLUSTER_PROVIDED` now
+	// Only apply auto mtls to clusters excluding these kind
+	if cluster.LbPolicy != apiv2.Cluster_CLUSTER_PROVIDED {
```
Comment (diemtvu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Can you add an unit test for this change?

Comment (incfly) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: +1.

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: ok

**Comment by hzxuzhonghu** [comment]
not completely, with autoMTLS, if the proxy and the service instances are all cluster internal, then we will set the tls mode to `TLSSettings_ISTIO_MUTUAL` in `conditionallyConvertToIstioMtls()` .

The only difference is this pr will not apply `TransportSocketMatches` for headless servcie. It set `TransportSocket` for headless.

**Comment by hzxuzhonghu** [comment]
@incfly any other comments?

**Comment by incfly** [comment]
If I understand correctly, so because of auto mTLS, in `conditionallyConvertToIstioMtls`, when customer don't have Destinationrule.TLS specified, we will set mode to be ISTIO_MUTUAL.

And then later on, in the `applyUpstreamTLSSettings`, around where this PR is modified, we'll set the TLS mode to be ISTIO_MUTUAL, so client and server are working under mTLS, just not using transport_socket_match.

If so, there's one issue, imagine some headless service like kafka, if 

- Does not have server workloads migrated with Istio sidecar (either no sidecar or partially sidecar)
- Define a namespace mTLs policy as PERMISSIVE. (in conditional convert, PERMISSIVE will make client tls mode as MUTUAL as well)

Then things will be break, right? i understand it's not working before either... (setting match on headless svc does not work)

in this situation, they need to set DR.TLS as disabled, or not having auto mTLS, or setting namespace policy as plaintext (effectively make it as Unknown).

Use case we support does not seem reduce.

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -1194,21 +1194,26 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.TLSSetting
...
-		transportSocket := cluster.TransportSocket
-		cluster.TransportSocket = nil
-		cluster.TransportSocketMatches = []*apiv2.Cluster_TransportSocketMatch{
-			{
-				Name: "tlsMode-" + model.IstioMutualTLSModeLabel,
-				Match: &structpb.Struct{
-					Fields: map[string]*structpb.Value{
-						model.TLSModeLabelShortname: {Kind: &structpb.Value_StringValue{StringValue: model.IstioMutualTLSModeLabel}},
+
+	// For `Cluster_ORIGINAL_DST` cluster discovery type, the LbPolicy is `Cluster_CLUSTER_PROVIDED` now
```
Comment (incfly) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: is there any param to refer cluster_discovery_type directly? will be nicer if we can use that, reduce one level indirection and coupling knowledge...
if not,

could we start with "for headless service, discover type will be original_dst, .and corresponding lbpolicy is CLUSTER_provided".

thus we know the source of motivation is from headless for future developer.

**Comment by incfly** [comment]
> If so, there's one issue, imagine some headless service like kafka, if

> Does not have server workloads migrated with Istio sidecar (either no sidecar or partially sidecar)
Define a namespace mTLs policy as PERMISSIVE. (in conditional convert, PERMISSIVE will make client tls mode as MUTUAL as well)
Then things will be break, right? i understand it's not working before either... (setting match on headless svc does not work)

> in this situation, they need to set DR.TLS as disabled, or not having auto mTLS, or setting namespace policy as plaintext (effectively make it as Unknown).

@rshriram this is touching the case where we discussed before. Server workloads can have partial sidecar migrated, but we now set client as `ISTIO_MUTUAL` due to a namespace policy or mesh policy (exists by default installation). They use Dr.TLS.Disabled to opt-out

I feel a slightly better approach is opposite. For headless, default to plaintext, requiring ISTIO_MUTUAL to opt-in.

wdty?

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 1 above)
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: For reviewers, this is the main change.

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: make sense

**Comment by hzxuzhonghu** [comment]
@incfly What you said is independent of this pr i think? It is definitely a config error?

**Comment by howardjohn** [comment]
@hzxuzhonghu this does not build

**Comment by hzxuzhonghu** [comment]
fixed, there was a pr updating the function

**Pull request merged**